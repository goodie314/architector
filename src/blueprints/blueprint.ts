import { BlueprintAttributes } from '../models/blueprint-attributes';
import ElementRef from './utils/element-ref';
import { EventHandler } from '../types/event-handler';
import BlueprintList from './blueprint-list';
import DynamicProp from './utils/dynamic-prop';

export class Blueprint {
    private readonly tagName: string;
    private plans: BlueprintAttributes;
    private elementRef: ElementRef<HTMLElement>;

    constructor(tagName: string) {
        this.tagName = tagName;
        this.plans = {
            classNames: [],
            attributes: {},
            children: [],
            handlers: {},
        };
    }

    id(value: string | DynamicProp<string>) {
        if (value instanceof DynamicProp) {
            value.onChange((val) => {
                this.plans.id = val
            });
        } else {
            this.plans.id = value;
        }
        return this;
    }

    classNames(...values: string[]) {
        this.plans.classNames.push(...values);
        return this;
    }

    attribute(name: string, value: string) {
        this.plans.attributes[name] = value;
        return this;
    }

    text(text: string) {
        this.plans.text = text;
        return this;
    }

    append(...components: (Blueprint | string | BlueprintList)[]) {
        components.forEach((component) =>
            this.plans.children.push(component),
        );
        return this;
    }

    ref(elementRef: ElementRef) {
        this.elementRef = elementRef;
        return this;
    }

    addEventListener(eventName: string, handler: EventHandler) {
        this.plans.handlers[eventName] = handler;
        return this;
    }

    click(eventHandler: EventHandler) {
        return this.addEventListener('click', eventHandler);
    }

    static build(describer: Blueprint) {
        const elem = document.createElement(describer.tagName);
        const describerAttributes = describer.plans;

        if (describerAttributes.id) {
            elem.id = describerAttributes.id;
        }
        if (describerAttributes.text) {
            elem.textContent = describerAttributes.text;
        }

        describerAttributes.classNames.forEach((className) =>
            elem.classList.add(className),
        );

        Object.entries(describerAttributes.attributes).forEach(([key, value]) =>
            elem.setAttribute(key, value),
        );

        describerAttributes.children
            .flatMap<string | HTMLElement>((child) => {
                if (child instanceof Blueprint) {
                    return [Blueprint.build(child)];
                } else if (child instanceof BlueprintList) {
                    return BlueprintList.Build(child, elem);
                } else {
                    return [child as string];
                }
            })
            .forEach((child) => elem.append(child));

        Object.entries(describerAttributes.handlers).forEach(
            ([eventName, eventHandler]) =>
                elem.addEventListener(eventName, (event) => {
                    if (!(event.currentTarget instanceof Element)) {
                        throw new Error(
                            'Internal error. It should not be possible to attach an event listener here where an element is not the target',
                        );
                    }
                    return eventHandler(event.currentTarget, event);
                }),
        );

        if (describer.elementRef) {
            describer.elementRef.set(elem);
        }

        return elem;
    }
}
