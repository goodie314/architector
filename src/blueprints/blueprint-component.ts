import { Blueprint } from './blueprint';
import BlueprintContext from '../structure/blueprint-context';

export abstract class BlueprintComponent {
    public context: BlueprintContext;

    constructor(componentName = 'default') {
        if (!this.context) {
            this.context = BlueprintContext.createContext(componentName);
        }
        if (componentName) {
            this.context.attachComponent(this, componentName);
        }
    }

    abstract compose(): Blueprint;

    // utility methods to generate blueprints
    protected blueprint(tagName: string) {
        return new Blueprint(tagName);
    }

    protected fragment(...blueprints: Blueprint[]) {
        return Blueprint.Fragment(...blueprints);
    }

    protected div() {
        return this.blueprint('div');
    }
    protected span() {
        return this.blueprint('span');
    }
    protected h(number: number) {
        return this.blueprint(`h${number}`);
    }
    protected ul() {
        return this.blueprint('ul');
    }
    protected li(text?: string) {
        return this.blueprint('li').text(text);
    }
    protected form() {
        return this.blueprint('form');
    }
    protected label(forName?: string) {
        const label = this.blueprint('label');
        if (forName) {
            label.attribute('for', forName);
        }
        return label;
    }
    protected button(text?: string) {
        return this.blueprint('button').text(text);
    }
    protected input(type?: string, name?: string) {
        const input = this.blueprint('input');
        if (type) {
            input.attribute('type', type);
        }
        if (name) {
            input.attribute('name', name);
        }
        return input;
    }
    protected checkbox() {
        return this.input('checkbox');
    }
}
