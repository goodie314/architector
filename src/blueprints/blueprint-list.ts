import { Blueprint } from './blueprint';
import ElementRef from './utils/element-ref';
import { ErrorMessages } from '../constants/error-messages';

export default class BlueprintList<K = string> {
    private elementMap: Map<K, HTMLElement>;
    private order: K[];
    private parentElement: ElementRef<HTMLElement>;

    constructor() {
        this.elementMap = new Map();
        this.order = [];
        this.parentElement = new ElementRef<HTMLElement>();
    }

    get length() {
        return this.order.length;
    }

    get(id: K) {
        return this.elementMap.get(id);
    }

    async push(id: K, describer: Blueprint) {
        if (this.elementMap.has(id)) {
            throw new Error(ErrorMessages.BlueprintList.duplicateId(id));
        }
        const elem = Blueprint.build(describer);

        this.elementMap.set(id, elem);
        this.order.push(id);

        const parent = await this.parentElement.get();
        parent.append(elem);
    }

    async replaceById(id: K, describer: Blueprint) {
        if (!this.elementMap.has(id)) {
            throw new Error(ErrorMessages.BlueprintList.replaceByIdError(id));
        }
        const old = this.elementMap.get(id);

        const elem = Blueprint.build(describer);
        this.elementMap.set(id, elem);

        old.replaceWith(elem);
        await this.parentElement.get();
    }

    async replaceByIndex(index: number, describer: Blueprint) {
        if (index < 0 || index >= this.order.length) {
            throw new Error(
                ErrorMessages.BlueprintList.replaceByIndexError(index),
            );
        }
        const id = this.order[index];
        const old = this.elementMap.get(id);

        const elem = Blueprint.build(describer);
        this.elementMap.set(id, elem);

        old.replaceWith(elem);
        await this.parentElement.get();
    }

    remove(id: K) {
        if (!this.elementMap.has(id)) {
            throw new Error(ErrorMessages.BlueprintList.removeError(id));
        }
        this.elementMap.get(id).remove();
        this.elementMap.delete(id);
        const index = this.order.indexOf(id);
        this.order = this.order.splice(index, 1);
        this.order = this.order.filter((oID) => oID !== id);
    }

    async set(
        descriptions: {
            id: K;
            describer: Blueprint;
        }[],
    ) {
        const updatedMap = new Map<K, HTMLElement>();
        const updatedOrder: K[] = [];
        for (
            let i = 0;
            i < Math.max(this.order.length, descriptions.length);
            i++
        ) {
            // old and new list are same length
            if (i < this.order.length && i < descriptions.length) {
                const oldId = this.order[i];
                const oldElem = this.elementMap.get(oldId);
                const { id, describer } = descriptions[i];
                const elem =
                    this.elementMap.get(id) ?? Blueprint.build(describer);
                updatedOrder.push(id);
                updatedMap.set(id, elem);
                if (oldId !== id) {
                    oldElem.replaceWith(elem);
                }
            }
            // old list is longer
            else if (i < this.order.length) {
                const oldId = this.order[i];
                const oldElem = this.elementMap.get(oldId);
                oldElem.remove();
            }
            // new list is longer
            else {
                const { id, describer } = descriptions[i];
                const elem =
                    this.elementMap.get(id) ?? Blueprint.build(describer);
                updatedOrder.push(id);
                updatedMap.set(id, elem);
                const parent = await this.parentElement.get();
                parent.append(elem);
            }
        }

        this.elementMap = updatedMap;
        this.order = updatedOrder;
    }

    static Build(
        describedList: BlueprintList<any>,
        parentElement: HTMLElement,
    ) {
        describedList.parentElement.set(parentElement);
        return describedList.order.map((id) =>
            describedList.elementMap.get(id),
        );
    }

    // possibly can't keep this method, can't think of a way to do it without
    // returning a promise which wouldn't be a very good flow for the user
    static from<K>(descriptions: { id: K; describer: Blueprint }[]) {
        const list = new BlueprintList<K>();
        descriptions.forEach((description) => {
            list.order.push(description.id);
            list.elementMap.set(
                description.id,
                Blueprint.build(description.describer),
            );
        });

        list.parentElement.get().then((parent) => {
            list.order.forEach((id) => parent.append(list.elementMap.get(id)));
        });

        return list;
    }
}
