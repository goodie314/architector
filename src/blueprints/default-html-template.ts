import BlueprintComponent from './blueprint-component';
import { Blueprint } from './blueprint';
import { BlueprintRenderable } from '../models/types';

export default class DefaultHtmlTemplate extends BlueprintComponent {
    private readonly title: string;
    private _language: string;
    private _externalStyles: string[];
    private _charset: string;
    private _viewport: string;
    private readonly _bodyContent: BlueprintRenderable[];

    constructor(title: string) {
        super();
        this.title = title;
        this._language = 'en';
        this._externalStyles = [];
        this._charset = 'UTF-8';
        this._viewport = 'width=device-width, initial-scale=1.0';
        this._bodyContent = [];
    }

    language(lang: string) {
        this._language = lang;
        return this;
    }

    externalStyle(url: string) {
        this._externalStyles.push(url);
        return this;
    }

    charset(_charset: string) {
        this._charset = _charset;
        return this;
    }

    viewport(_viewport: string) {
        this._viewport = _viewport;
        return this;
    }

    bodyElement(bodyElem: BlueprintRenderable) {
        this._bodyContent.push(bodyElem);
        return this;
    }

    compose(): Blueprint {
        const styles = this._externalStyles.map((url) =>
            this.blueprint('style').attribute('href', url),
        );

        return Blueprint.Fragment(
            this.blueprint('!DOCTYPE').attribute('HTML'),
            this.blueprint('html')
                .attribute('lang', this._language)
                .append(
                    this.blueprint('head').append(
                        this.blueprint('meta').attribute(
                            'charset',
                            this._charset,
                        ),
                        this.blueprint('meta')
                            .attribute('name', 'viewport')
                            .attribute('content', this._viewport),
                        this.blueprint('title').text(this.title),
                        ...styles,
                    ),
                    this.blueprint('body').append(...this._bodyContent),
                ),
        );
    }
}
