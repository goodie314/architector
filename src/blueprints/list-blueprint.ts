import BlueprintComponent from './blueprint-component';
import { Blueprint } from './blueprint';
import DynamicProp from './utils/dynamic-prop';
import ElementRef from './utils/element-ref';
import BlueprintBuilder from '../structure/blueprint-builder';
import { compareValues } from '../util/value-util';

type BlueprintMapper<T> = (dataElement: T) => Blueprint;
type IdMapper<T> = (dataElement: T) => string;

export default class ListBlueprint<T = any> extends BlueprintComponent {
    private readonly blueprintMapper: BlueprintMapper<T>;
    private readonly idMapper: IdMapper<T>;
    private containerRef = new ElementRef();

    private order: string[];
    private elementMap: Map<string, HTMLElement>;

    constructor(
        data: DynamicProp<T[]>,
        blueprintMapper: BlueprintMapper<T>,
        idMapper?: IdMapper<T>,
    ) {
        super();
        this.blueprintMapper = blueprintMapper;
        this.idMapper =
            idMapper ??
            ((dataElement) => {
                // @ts-ignore
                if (dataElement['id'] == null) {
                    throw new Error(
                        `Data elements do not have an id prop. You must pass in a mapping function to produce an id`,
                    );
                }

                // @ts-ignore
                return dataElement['id'];
            });

        this.order = [];
        this.elementMap = new Map();

        data.onElementChange(
            this.containerRef,
            this.handleElementReady.bind(this),
        );
    }

    private handleElementReady(element: HTMLElement) {
        return (values: T[], prevValues: T[]) => {
            const safeGetOrder = (index: number) =>
                index < this.order.length ? this.order[index] : null;

            const updatedOrder = values.map(this.idMapper);
            const updatedMap = new Map<string, HTMLElement>();
            const updatedElements: HTMLElement[] = [];
            const prevValueMap = new Map<string, T>();
            prevValues.forEach((val) =>
                prevValueMap.set(this.idMapper(val), val),
            );

            for (let i = 0; i < updatedOrder.length; i++) {
                const id = updatedOrder[i];
                const oldId = safeGetOrder(i);
                const data = values[i];
                if (id !== oldId) {
                    if (this.elementMap.has(id)) {
                        const updatedElement = compareValues(
                            data,
                            prevValueMap.get(id),
                        )
                            ? this.elementMap.get(id)
                            : BlueprintBuilder.build(
                                  this.blueprintMapper(data),
                              );
                        updatedMap.set(id, updatedElement);
                        updatedElements.push(updatedElement);
                    } else {
                        const updatedElement = BlueprintBuilder.build(
                            this.blueprintMapper(data),
                        );
                        updatedMap.set(id, updatedElement);
                        updatedElements.push(updatedElement);
                    }
                } else {
                    const updatedElement = compareValues(
                        data,
                        prevValueMap.get(id),
                    )
                        ? this.elementMap.get(id)
                        : BlueprintBuilder.build(this.blueprintMapper(data));
                    updatedMap.set(id, updatedElement);
                    updatedElements.push(updatedElement);
                }
            }

            this.order
                .filter((id) => !updatedMap.has(id))
                .map((id) => this.elementMap.get(id))
                .forEach((el) => el.remove());
            this.order = updatedOrder;
            this.elementMap = updatedMap;

            element.innerHTML = '';
            updatedElements.forEach((child) => element.append(child));
        };
    }
    compose(): Blueprint {
        return this.fragment().ref(this.containerRef);
    }
}
