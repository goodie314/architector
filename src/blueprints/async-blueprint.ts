import { BlueprintComponent } from './blueprint-component';
import { Blueprint } from './blueprint';
import { BlueprintCallback } from '../models/types';
import ElementRef from './utils/element-ref';
import { BlueprintBuilder } from '../structure/blueprint-builder';

export default class AsyncBlueprint extends BlueprintComponent {
    private promise: Promise<any>;
    private _waitingView: Blueprint;
    private _resolvedView: BlueprintCallback;
    private _rejectedView: BlueprintCallback;
    private renderRef = new ElementRef();

    constructor(promise: Promise<any>) {
        super();
        this.promise = promise;
        this._waitingView = Blueprint.Fragment();
    }

    waitingView(blueprint: Blueprint) {
        this._waitingView = blueprint;
        return this;
    }

    resolvedView(callback: BlueprintCallback) {
        this._resolvedView = callback;
        return this;
    }

    rejectedView(callback: BlueprintCallback) {
        this._rejectedView = callback;
        return this;
    }

    compose(): Blueprint {
        this.promise
            .then((value) =>
                this.renderRef.get().then((elem) => {
                    elem.innerHTML = '';
                    const updatedView = BlueprintBuilder.build(
                        this._resolvedView(value),
                    );
                    elem.append(updatedView);
                }),
            )
            .catch((err) =>
                this.renderRef.get().then((elem) => {
                    if (this._rejectedView) {
                        elem.innerHTML = '';
                        const updatedView = BlueprintBuilder.build(
                            this._rejectedView(err),
                        );
                        elem.append(updatedView);
                    }
                }),
            );

        return this._waitingView.ref(this.renderRef);
    }
}
