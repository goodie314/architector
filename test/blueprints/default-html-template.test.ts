import BlueprintHTMLBuilder from '../../src/structure/blueprint-html-builder';
import { DefaultHtmlTemplate } from '../../src/blueprints/default-html-template';

describe('DefaultHtmlTemplate', () => {
    test('renders default template', () => {
        expect(
            BlueprintHTMLBuilder(new DefaultHtmlTemplate('test title')),
        ).toMatchSnapshot();
    });
});
