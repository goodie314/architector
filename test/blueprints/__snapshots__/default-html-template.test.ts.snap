// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultHtmlTemplate renders default template 1`] = `
"<!DOCTYPE HTML>
<html lang="en">
	<head>
		<meta charset="UTF-8"/>
		<meta name="viewport" content="width=device-width, initial-scale=1.0"/>
		<title>
			test title
		</title>
	</head>
	<body>
	</body>
</html>"
`;
